src/librustc_typeck/check/expr.rs

```diff
@@ -1541,12 +1541,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
-    fn check_expr_yield(&self, value: &'tcx hir::Expr, expr: &'tcx hir::Expr) -> Ty<'tcx> {
+    fn check_expr_yield(
+        &self,
+        value: &'tcx hir::Expr,
+        expr: &'tcx hir::Expr,
+        src: &'tcx hir::YieldSource
+    ) -> Ty<'tcx> {
         match self.yield_ty {
             Some(ty) => {
                 self.check_expr_coercable_to_type(&value, ty);
             }
-            None => {
+            // Given that this `yield` expression was generated as a result of lowering a `.await`,
+            // we know that the yield type must be `()`; however, the context won't contain this
+            // information. Hence, we check the source of the yield expression here and check its
+            // value's type against `()` (this check should always hold).
+            None if src == &hir::YieldSource::Await => {
```

Comment:
I wonder if the better fix is in `check_fn` instead, namely from:
https://github.com/rust-lang/rust/blob/d3e2cec29225a46298ec4ebf082f34ebd7cfeecf/src/librustc_typeck/check/mod.rs#L1073-L1080
into:
```rust
    if body.generator_kind.is_some() && can_be_generator.is_some() {
        let yield_ty = match body.generator_kind {
            GeneratorKind::Async => fcx.tcx.mk_unit(),
            GeneratorKind::Gen => fcx.next_ty_var(TypeVariableOrigin {
                kind: TypeVariableOriginKind::TypeInference,
                span,
            }),
        };
        fcx.require_type_is_sized(yield_ty, span, traits::SizedYieldType);
        fcx.yield_ty = Some(yield_ty);
    }
```

Reply:
I don't think this avoids the need for the special case as `cam_be_generator.is_some()` won't hold in the contexts in which we were getting an ICE (i.e. bodies that weren't `async`). The need for the special case arises because we are type checking the sub-expression of a `.await` expression despite the fact that the whole `.await` expression shouldn't have been there in the first place.

Reply:
I don't think this avoids the need for the special case as `cam_be_generator.is_some()` won't hold in the contexts in which we were getting an ICE (i.e. bodies that weren't `async`). The need for the special case arises because we are type checking the sub-expression of a `.await` expression despite the fact that the whole `.await` expression shouldn't have been there in the first place.

Reply:
Ah yeah, that's true.